src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs

```diff
@@ -9,6 +11,128 @@ namespace System.Collections.Immutable.Test
     public class ImmutableInterlockedTests
     {
         [Fact]
+        public void ApplyChange_StartWithNull()
+        {
+            ImmutableList<int> list = null;
+            Assert.True(ImmutableInterlocked.ApplyChange(ref list, l => { Assert.Null(l); return ImmutableList.Create(1); }));
+            Assert.Equal(1, list.Count);
+            Assert.Equal(1, list[0]);
+        }
+
+        [Fact]
+        public void ApplyChange_IncrementalUpdate()
+        {
+            ImmutableList<int> list = ImmutableList.Create(1);
+            Assert.True(ImmutableInterlocked.ApplyChange(ref list, l => l.Add(2)));
+            Assert.Equal(2, list.Count);
+            Assert.Equal(1, list[0]);
+            Assert.Equal(2, list[1]);
+        }
+
+        [Fact]
+        public void ApplyChange_FuncThrowsThrough()
+        {
+            ImmutableList<int> list = ImmutableList.Create(1);
+            Assert.Throws<InvalidOperationException>(() => ImmutableInterlocked.ApplyChange(ref list, l => { throw new InvalidOperationException(); }));
+        }
+
+        [Fact]
+        public void ApplyChange_NoEffectualChange()
+        {
+            ImmutableList<int> list = ImmutableList.Create<int>(1);
+            Assert.False(ImmutableInterlocked.ApplyChange(ref list, l => l));
+        }
+
+        [Fact]
+        public void ApplyChange_HighConcurrency()
+        {
+            ImmutableList<int> list = ImmutableList.Create<int>();
+            int concurrencyLevel = Environment.ProcessorCount;
+            int iterations = 500;
+            Task[] tasks = new Task[concurrencyLevel];
+            var countdown = new CountdownEvent(tasks.Length);
```

Comment:
Minor, but you can use a Barrier instead of a CountdownEvent, since that's basically the semantics you're trying to achieve.